ios/engine/KMEI/KeymanEngine/Classes/KeymanTextField.swift

```diff
@@ -0,0 +1,355 @@
+      return UInt16(codeUnit)
+    }
+
+    let text = String(utf16CodeUnits: codeUnits, count: codeUnits.count)
+
+    let textRange = selectedTextRange!
+    let selRange = NSRange(location: offset(from: beginningOfDocument, to: textRange.start),
+                           length: offset(from: textRange.start, to: textRange.end))
+
+    if dn <= 0 {
+      if selRange.length == 0 {
+        insertText(text)
+      } else {
+        self.text = (self.text! as NSString).replacingCharacters(in: selRange, with: text)
+      }
+    } else {
+      if s.isEmpty {
+        for _ in 0..<dn {
+          deleteBackward()
+        }
+      } else {
+        if selRange.length == 0 {
+          for _ in 0..<dn {
+            deleteBackward()
+          }
+          insertText(text)
+        } else {
+          self.text = (self.text! as NSString).replacingCharacters(in: selRange, with: text)
+        }
+      }
+    }
+  }
+
+  // MARK: - UITextFieldDelegate Hooks
+  public override var selectedTextRange: UITextRange? {
+    didSet(range) {
+      guard let range = range else {
+        return
+      }
+      let newRange = NSRange(location: offset(from: beginningOfDocument, to: range.start),
+                             length: offset(from: range.start, to: range.end))
+      Manager.shared.setSelectionRange(newRange, manually: false)
+    }
+  }
+
+  public func textField(_ textField: UITextField, shouldChangeCharactersIn range: NSRange,
+                        replacementString string: String) -> Bool {
+    shouldUpdateKMText = true // Enable text update to catch copy/paste operations
+    return true
+  }
+
+  @objc public func textFieldTextDidChange(_ notification: Notification) {
+    if shouldUpdateKMText {
+      // Catches copy/paste operations
+      Manager.shared.setText(text)
+      let textRange = selectedTextRange!
+      let newRange = NSRange(location: offset(from: beginningOfDocument, to: textRange.start),
+                             length: offset(from: textRange.start, to: textRange.end))
+      Manager.shared.setSelectionRange(newRange, manually: false)
+      shouldUpdateKMText = false
+    }
+  }
+
+  public func textFieldShouldClear(_ textField: UITextField) -> Bool {
+    if textField == self {
+      Manager.shared.clearText()
+    }
+    return true
+  }
+
+  public func textFieldShouldBeginEditing(_ textField: UITextField) -> Bool {
+    let leftToRightMark = "\u{200e}"
+    let rightToLeftOverride = "\u{202e}"
+    let textWD = baseWritingDirection(for: beginningOfDocument, in: .forward)
+
+    let isRTL: Bool
+    if let keyboardID = Manager.shared.keyboardID,
+      let languageID = Manager.shared.languageID {
+      isRTL = Manager.shared.isRTLKeyboard(withID: keyboardID, languageID: languageID)
+    } else {
+      isRTL = false
+    }
+
+    if isRTL {
+      if textWD != .rightToLeft {
+        if text.hasPrefix(leftToRightMark) {
+          text = String(text.utf16.dropFirst())
```

Comment:
I take it that `text.utf16.dropFirst()` replaces [this linked block](https://github.com/keymanapp/keyman/pull/378/files#diff-b6ef7f3c4df91bbb8a742e9a09330e20L247) (within the if)?

Reply:
Yep that's the intent.